### PR TITLE
Allow uninstall from symlinked GEM_HOME

### LINKED ISSUE #2705: gem uninstall -i path doesn't work if path is a symlink
```
$ gem env                                                                     
RubyGems Environment:
  - RUBYGEMS VERSION: 3.0.3
  - RUBY VERSION: 2.6.2 (2019-03-13 patchlevel 47) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/eregon/.gem/ruby/2.6.2
  - USER INSTALLATION DIRECTORY: /home/eregon/.gem/ruby/2.6.0
  - RUBY EXECUTABLE: /home/eregon/.rubies/ruby-2.6.2/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /home/eregon/.gem/ruby/2.6.2/bin
  - SPEC CACHE DIRECTORY: /home/eregon/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /home/eregon/.rubies/ruby-2.6.2/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /home/eregon/.gem/ruby/2.6.2
     - /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "gem" => "--no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /home/eregon/.gem/ruby/2.6.2/bin
     - /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0/bin
     - /home/eregon/.rubies/ruby-2.6.2/bin
     - /home/eregon/bin
     - /usr/local/bin
     - /usr/local/sbin
     - /usr/bin
     - /usr/sbin
     - /home/eregon/code/mx
     - /home/eregon/code/mspec/bin
     - /opt/crystal/bin
$ gem install --install-dir /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0 path
Successfully installed path-2.0.1
1 gem installed
$ ln -s /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0 sym
$ gem uninstall -i sym path
# No output, and the gem is still there
$ gem uninstall -i /home/eregon/.rubies/ruby-2.6.2/lib/ruby/gems/2.6.0 path
# Works
```

This notably happens if one gives the `..@global` directory for removing gems when using RVM, such as what the CI does currently:
https://github.com/rubygems/rubygems/blob/78022d6a1136007dab6d8e4dadc8b0f229d0158c/util/ci#L52

Related to #2701.

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
$ gem env version
3.0.3
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

Relates to #2700.

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
# Description:

Currently, if GEM_HOME is a symlink (for example, in rvm's ruby-head), `gem uninstall` fails. I hope to fix that in this PR.

Fixes #2705.

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
It'd be great if someone with a Windows machine, or with some Windows knowledge could have a look at Appveyor :sweat_smile:.

File: test/rubygems/test_gem_uninstaller.rb [hunk 1]
```diff
@@ -149,6 +151,20 @@ def test_remove_not_in_home
     assert_path_exists @spec.gem_dir
   end
 
+  def test_remove_symlinked_gem_home
+    symlinked_gem_home = Tempfile.new("gem_home").path
+
+    FileUtils.ln_sf(@gemhome, symlinked_gem_home)
```
Comment (p-mongo) [inline test/rubygems/test_gem_uninstaller.rb]: I imagine at this point the file created by Tempfile is still open, and this line tries to remove it and replace it with a symlink. This works on unix but not on windows.

My suggestion is to create a temporary directory instead of a temporary file and use `ln_s` instead of `ln_sf` to create a symlink within that directory.

**Comment by MSP-Greg** [comment]
@deivid-rodriguez 

> It'd be great if someone with a Windows machine, or with some Windows knowledge

1st - yes, 2nd - questionable...

Windows won't make a symlink from an existing file.  Something like the following:

```ruby
  def test_remove_symlinked_gem_home
    symlinked_gem_home = (tmp = Tempfile.new("gem_home")).path
    tmp.close(false) if windows?            # windows won't close an open file

    FileUtils.ln_sf(@gemhome, symlinked_gem_home)
```

**Comment by p-mongo** [comment]
That pattern is a temporary file race (see https://dwheeler.com/secure-programs/Secure-Programs-HOWTO/avoid-race.html section 7.11.1.2. Temporary Files).

**Comment by MSP-Greg** [comment]
Well, if the intent is to get a safe name, not many other options.  I ran it in Appveyor and it passed.  Ruby Core runs tests parallel, I haven't tried it there.

Since Windows will not allow one to create a symlink from an existing file system name, and `Tempfile.new` returns a object that should be open, so I'm not sure I see the problem...

**Comment by p-mongo** [comment]
Creating a temporary directory, and working within it, is the safe alternative.

https://ruby-doc.org/stdlib-1.9.3/libdoc/tmpdir/rdoc/Dir.html#method-c-mktmpdir

**Comment by deivid-rodriguez** [comment]
Thanks both for your help! It should be fixed now.

**Comment by deivid-rodriguez** [comment]
As can be seen in the tests I had to modify, the uninstaller now complains if you pass it a non-existent directory. Sounds like a good thing.

**Comment by deivid-rodriguez** [comment]
I also took the chance to fix our CI against ruby-head, since it was hitting precisely this bug as pointed out by @eregon in #2705.

**Comment by deivid-rodriguez** [comment]
Can we get this in for a fully green CI matrix, including ruby-head? :smiley:

File: util/ci [hunk 2]
```diff
@@ -49,7 +49,7 @@ end
 case ARGV
 when %w(before_script)
   if TOOL.rubygems?
-    run('gem', %W(uninstall executable-hooks gem-wrappers bundler-unload -x --force -i #{`gem env home`.strip}@global))
+    run('ruby', %W(-I lib bin/gem uninstall executable-hooks gem-wrappers bundler-unload -x --force -i #{`gem env home`.strip}@global))
```
Comment (bronzdoc) [inline util/ci]: Why is this change needed here?

Comment (deivid-rodriguez) [reply util/ci]: Not really "needed" but our build against `ruby-head` is [hitting precisely this bug](https://travis-ci.org/rubygems/rubygems/jobs/519464636#L236-L241). So by changing this, we get to use this fix, and thus get it green.

Comment (bronzdoc) [reply util/ci]: Maybe have this in a separate PR? just so this change doesn't get obscure in the purpose of this PR, what you think?

Comment (deivid-rodriguez) [reply util/ci]: Sure.

Comment (deivid-rodriguez) [reply util/ci]: I removed the commit fixing the ruby-head build and will open a separate PR once this one is merged.

Comment (bronzdoc) [reply util/ci]: Maybe we can merge the other one first and then rebase this one? so this is green in ruby-head too :)

Comment (deivid-rodriguez) [reply util/ci]: We need the fix in this PR as a prerequisite to fix ruby-head!

Comment (deivid-rodriguez) [reply util/ci]: If we want this PR to be green in ruby-head, I need to restore the commit I removed.

Comment (deivid-rodriguez) [reply util/ci]: That's why I put both commits in this PR, because they are very related.

**Comment by bronzdoc** [comment]
@bundlerbot r+
